contracts/gol.cairo

```diff
@@ -0,0 +1,450 @@
+func ensure_user{
+        syscall_ptr : felt*
+    }(
+    ) -> (
+        caller : felt
+    ):
+    let (caller) = get_caller_address()
+    with_attr error_message(
+        "User not authenticated"
+    ):
+        assert_not_zero(caller)
+    end
+    return (caller)
+end
+
+func verify_new_game{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        game : felt
+    ):
+    alloc_locals
+    let (local game_info) = stored_game.read(
+        game_id=game, generation=0)
+    let game_id = game_info[1]
+
+    with_attr error_message(
+        "Game with genesis state {game} already exists"
+    ):
+        assert_not_equal(game_id, game)
+    end
+
+    let (high, _) = split_felt(game)
+    let (invalid_bits, _) = unsigned_div_rem(high, 2**97)
+    with_attr error_message(
+        "Game {game} is not a valid game"
+    ):
+        assert invalid_bits = 0
+    end
+    return ()
+end
+
+func verify_and_get_for_cell_activation{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        cell_index : felt,
+        user : felt
+    ) -> (
+        current_generation : felt,
+        current_state : felt
+    ):
+    alloc_locals
+    let (generation) = current_generation.read(
+        game_id=INFINITE_GAME_GENESIS
+    )
+    let (local game_info) = stored_game.read(
+        game_id=INFINITE_GAME_GENESIS,
+        generation=generation
+    )
+    let current_game_state = game_info[1]
+    with_attr error_message(
+        "Cell index {cell_index} out of range"
+    ):
+        assert_le_felt(cell_index, DIM*DIM-1)
+    end
+    return(generation, current_game_state)
+end
+
+func save_game{
+        syscall_ptr : felt*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        generation : felt,
+        game_state : felt,
+        user_id : felt
+    ):
+    stored_game.write(
+        game_id=game_state,
+        generation=generation,
+        value=(user_id, game_state)
+    )
+    current_generation.write(game_state, generation)
+    game_created.emit(
+        owner_id=user_id,
+        game_id=game_state,
+        state=game_state
+    )
+    return ()
+end
+
+# User input may override state to make a cell alive.
+func activate_cell{
+        syscall_ptr : felt*,
+        bitwise_ptr : BitwiseBuiltin*,
+        pedersen_ptr : HashBuiltin*,
+        range_check_ptr
+    }(
+        generation : felt,
+        caller : felt,
+        cell_index : felt,
+        current_state : felt
+    ):
+    alloc_locals
+    let (local updated_cells : felt*) = alloc()
+
+    local upper_len = DIM*DIM - 1 - cell_index
+    local offset = cell_index + 1
+
+    let (cells_len, cells) = unpack_game(
+        game=current_state
+    )
+
+    memcpy(updated_cells, cells, cell_index)
+    assert updated_cells[cell_index] = 1
+    memcpy(updated_cells + offset, cells + offset, upper_len)    
```

Comment:
A cheaper way of doing this would be to do something like this:
```
let enabled_bit = 2**index
let new_board = bitwise_or(board, enabled_board)
```
This way we avoid unpacking , copying and packing it again for each cell. There are 225 cells, so at least 675 operations have to be done (there's more for sure) - that's much more than a single bitwise https://docs.starknet.io/docs/Fees/fee-mechanism/.